### PR TITLE
delta: support instantiated units and dropins

### PR DESCRIPTION
**Pull request by jsynacek**
Consider the following filesystem contents:

```
  /usr/lib/systemd/system/httpd.service
    env.conf
  /usr/lib/systemd/system/postfix.service
  /usr/lib/systemd/system/getty@.service
  /usr/lib/systemd/system/getty@.service.d/
    ext.conf
  /etc/systemd/system/getty@tty1.service.d/
    test1-a.conf
    test1-b.conf
  /etc/systemd/system/getty@tty2.service
    test2-c.conf
    test2-d.conf
  /etc/systemd/system/getty@tty2.service.d/
  /run/systemd/system/getty@tty5.service
```

The resulting internal representiation will be as follows.

```
   top[httpd.service] = /usr/lib/systemd/system/httpd.service
   bottom[httpd.service]
     [/usr/lib/systemd/system/httpd.service] -> env.conf

   top[postfix.service] = /usr/lib/systemd/system/postfix.service
   bottom[postfix.service]
     [/usr/lib/systemd/system/postfix.service] -> (empty list)

   top[getty@.service] = /usr/lib/systemd/system/getty@.service
   bottom[getty@.service]
     [/usr/lib/systemd/system/getty@.service]  -> ext.conf
     [/etc/systemd/system/getty@tty1.service]  -> test1-a.conf, test1-b.conf
     [/etc/systemd/system/getty@tty2.service]  -> test2-c.conf, test2-d.conf
     [/run/systemd/systemd/getty@tty5.service] -> (empty list)
```

The 'top' hashmap contains the filename of the very first unit, which is
considered as the basis for extension. It's key is a base name of the
full path of the unit it holds as its value, or basename of an template
unit.

The 'bottom' hashmap holds a hashmap with override-to-extension
mappings and uses the same key as the 'top' hashmap. The extensions are
saved in a plain string list. Empty list means that there are only
overrides to the corresponding unit in 'top'.

### CODE REVIEW

File: src/delta/delta.c [hunk 1]
```diff
@@ -207,102 +208,95 @@ static int found_override(const char *top, const char *bottom) {
...
+                        k = hashmap_put(drops, path, NULL);
+                        if (k < 0)
+                                return k;
                 }
 
-                p = strdup(p);
-                if (!p)
+                d = hashmap_get(drops, path);
+                k = strv_extend(&d, *file);
+                if (k < 0)
```
Comment (zonque) [inline src/delta/delta.c:272]: You should rather return k directly here, instead of making up an error code.

File: src/delta/delta.c [hunk 2]
```diff
@@ -207,102 +208,95 @@ static int found_override(const char *top, const char *bottom) {
...
+
+                key = strdup(key);
+                if (!key)
+                        return -ENOMEM;
+
+                tmp = NULL;
+
+                path = strdup(path);
+                if (!path) {
+                        return -ENOMEM;
```
Comment (zonque) [inline src/delta/delta.c:285]: Curly braces are not needed for single-statement blocks.

File: src/delta/delta.c [hunk 3]
```diff
@@ -328,47 +324,94 @@ static int enumerate_dir(Hashmap *top, Hashmap *bottom, Hashmap *drops, const ch
...
+                key = tmp ?: basename(p);
+
+                log_debug("Adding: top[%s] %s %s", key, draw_special_char(DRAW_ARROW), p);
+                k = hashmap_put(top, key, p);
+
                 if (k < 0) {
-                        free(p);
-                        return k;
+                        if (k == -EEXIST) {
+                                p = strdup(p);
```
Comment (zonque) [inline src/delta/delta.c:348]: Here, you're losing the reference to the pointer returned by `strdup()` above, and hence leak `p`.

File: src/delta/delta.c [hunk 4]
```diff
@@ -328,47 +324,94 @@ static int enumerate_dir(Hashmap *top, Hashmap *bottom, Hashmap *drops, const ch
...
+
+                                        log_debug("Adding: bottom[%s] %s drops[%s] %s []",
+                                                  key, draw_special_char(DRAW_ARROW),
+                                                  p, draw_special_char(DRAW_ARROW));
+
+                                        k = hashmap_put(drops, p, NULL);
+                                        if (k < 0)
+                                                return k;
+
+                                        key = strdup(key);
```
Comment (zonque) [inline src/delta/delta.c:372]: Same here. `key` holds what was returned by `strdup(tmp);` above, and the pointer to this is now leaked. Try running valgrind on that tool.

I think you need to refactor that code a bit so the life time of the allocated strings becomes more obvious.

**Comment by zonque** [comment]
This version looks much better now, but as I commented inline, the life time of the allocated strings is hard to follow, and the code also leaks memory. You are right it does not matter much for such a standalone tool, but I also agree to @poettering here that we should maintain the same code quality throughout the entire code base.

**Comment by jsynacek** [comment]
V4: https://github.com/systemd/systemd/pull/1405
